bigtable/client/metadata_update_policy_test.cc

```diff
@@ -13,13 +13,278 @@
 // limitations under the License.
 
 #include "bigtable/client/metadata_update_policy.h"
+#include "bigtable/client/admin_client.h"
+#include "bigtable/client/internal/make_unique.h"
+#include "bigtable/client/table.h"
+#include "bigtable/client/table_admin.h"
+#include "google/bigtable/v2/bigtable.grpc.pb.h"
+#include <google/bigtable/admin/v2/bigtable_table_admin.grpc.pb.h>
 #include <gtest/gtest.h>
+#include <thread>
 
 std::string const kInstanceName = "projects/foo-project/instances/bar-instance";
 std::string const kTableId = "baz-table";
 std::string const kTableName =
     "projects/foo-project/instances/bar-instance/tables/baz-table";
 
+namespace btproto = google::bigtable::v2;
+namespace adminproto = google::bigtable::admin::v2;
+
+/**
+ * Implement the portions of the `google.bigtable.v2.Bigtable` interface
+ * necessary for the embedded server tests.
+ *
+ * This is not a Mock (use `google::bigtable::v2::MockBigtableStub` for that,
+ * nor is this a Fake implementation (use the Cloud Bigtable Emulator for that),
+ * this is an implementation of the interface that returns hardcoded values.
+ * It is suitable for the embedded server tests, but for nothing else.
+ */
+class BigtableImpl final : public btproto::Bigtable::Service {
+ public:
+  BigtableImpl() {}
+  grpc::Status ReadRows(
+      grpc::ServerContext* context,
+      google::bigtable::v2::ReadRowsRequest const* request,
+      grpc::ServerWriter<google::bigtable::v2::ReadRowsResponse>* writer)
+      override {
+    for (std::multimap<grpc::string_ref, grpc::string_ref>::const_iterator it =
+             context->client_metadata().begin();
+         it != context->client_metadata().end(); ++it) {
+      auto ele = std::make_pair((*it).first, (*it).second);
+      client_metadata_.emplace(ele);
+    }
+    return grpc::Status::OK;
+  }
+
+  const std::multimap<grpc::string_ref, grpc::string_ref>& client_metadata()
+      const {
+    return client_metadata_;
+  }
+
+ private:
+  std::multimap<grpc::string_ref, grpc::string_ref> client_metadata_;
+};
+
+/**
+ * Implement the `google.bigtable.admin.v2.BigtableTableAdmin` interface for the
+ * embedded server tests.
+ */
+class TableAdminImpl final : public adminproto::BigtableTableAdmin::Service {
+ public:
+  TableAdminImpl() {}
+
+  grpc::Status CreateTable(
+      grpc::ServerContext* context,
+      google::bigtable::admin::v2::CreateTableRequest const* request,
+      google::bigtable::admin::v2::Table* response) override {
+    for (std::multimap<grpc::string_ref, grpc::string_ref>::const_iterator it =
```

Comment:
Should we refactor this code to a function?  It appears at least 3 times here.

Reply:
done that in embedded_server_fixture_Test